lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureOutputStream.java

```diff
@@ -135,32 +129,106 @@ private void ensureOpen()
         }
     }
 
-    @Override
-    public void close()
-            throws IOException
+    private void ensureCapacity(int extra)
     {
-        if (!closed) {
-            closed = true;
-            try {
-                stream.close();
+        int capacity = min(writeBlockSizeBytes, bufferSize + extra);
+        if (buffer.length < capacity) {
+            int target = max(buffer.length, initialBufferSize);
+            if (target < capacity) {
+                target += target / 2; // increase 50%
+                target = clamp(target, capacity, writeBlockSizeBytes);
             }
-            catch (IOException e) {
-                // Azure close sometimes rethrows IOExceptions from worker threads, so the
-                // stack traces are disconnected from this call. Wrapping here solves that problem.
-                throw new IOException("Error closing file: " + location, e);
+            buffer = Arrays.copyOf(buffer, target);
+            memoryContext.setBytes(buffer.length);
+        }
+    }
+
+    private void flushBuffer(boolean finished)
+            throws IOException
+    {
+        // skip multipart upload if there would only be one staged block
+        if (finished && !multiStagingStarted) {
+            BlobOutputStream blobOutputStream = blockClient
+                    .getBlobOutputStream(overwrite);
+
+            blobOutputStream.write(buffer, 0, bufferSize);
+            blobOutputStream.close();
+            return;
+        }
+
+        if ((bufferSize == writeBlockSizeBytes) || (finished && (bufferSize > 0))) {
+            byte[] data = buffer;
+            int length = bufferSize;
+
+            if (finished) {
+                this.buffer = null;
             }
-            finally {
-                memoryContext.close();
+            else {
+                this.buffer = new byte[0];
+                this.initialBufferSize = writeBlockSizeBytes;
+                bufferSize = 0;
             }
+            memoryContext.setBytes(0);
+            waitForPreviousUploadFinish();
```

Comment:
You might want to have a separate context for the inprogress upload, that gets set to zero when the upload finishes. Then you can just swap current into inprogress after this line.